pkg/logql/engine.go

```diff
@@ -263,16 +259,11 @@ func (q *query) Eval(ctx context.Context) (promql_parser.Value, error) {
 	ctx, cancel := context.WithTimeout(ctx, queryTimeout)
 	defer cancel()
 
-	expr, err := q.parse(ctx, q.params.Query())
-	if err != nil {
-		return nil, err
-	}
-
 	if q.checkBlocked(ctx, tenants) {
 		return nil, logqlmodel.ErrBlocked
 	}
 
-	switch e := expr.(type) {
+	switch e := q.params.GetExpression().(type) {
```

Comment:
should we have the `Expr` interface also support a `Type` function? or maybe even via the `query` struct modify it to save the parsed `Expr` instead of just having `Params` and a `parse` function?

Reply:
We could but it would be very similar to this type switch. 

This is btw the benefit of the change. We don't have to parse the query.